src/lib.rs

```diff
@@ -9,84 +9,129 @@
+//! thread with thread-local storage.
+//! 
+//! If one wants to obtain random data directly from an external source it is
+//! recommended to use `EntropyRng` which manages multiple available sources
+//! or `OsRng` which retrieves random data directly from the OS. It should be
+//! noted that this is significantly slower than using a local generator like
+//! `thread_rng` and potentially much slower if `EntropyRng` must fall back to
+//! `JitterRng` as a source.
+//! 
+//! It is also common to use an algorithmic generator in local memory; for
+//! example `StdRng` or `SmallRng` (the main advantages of the latter are
+//! much smaller state and faster initialisation while its disadvantage is
+//! that it may be easy for an attacker to predict; actual performance varies).
+//! These can be seeded from a parent generator (`SeedableRng::thread_rng`) or
+//! with fresh entropy using `NewRng`:
+//! 
 //! ```
-//!
-//! and this to your crate root:
-//!
-//! ```rust
-//! extern crate rand;
+//! use rand::{SmallRng, NewRng};
+//! let mut rng = SmallRng::new();
 //! ```
-//!
-//! # Thread-local RNG
-//!
-//! There is built-in support for a RNG associated with each thread stored
-//! in thread-local storage. This RNG can be accessed via `thread_rng`, or
-//! used implicitly via `random`. This RNG is normally randomly seeded
-//! from an operating-system source of randomness, e.g. `/dev/urandom` on
-//! Unix systems, and will automatically reseed itself from this source
-//! after generating 32 KiB of random data.
-//!
+//! 
+//! In case one wants specifically to have a reproducible stream of "random"
+//! data (e.g. to procedurally generate a game world), select a named algorithm
+//! (i.e. not `StdRng`/`SmallRng` which may be adjusted in the future), and
+//! use `SeedableRng::from_seed`. Some generators provide additional
+//! constructors suitable for some applications, e.g. `IsaacRng::new_from_u64`.
+//! 
+//! # Applying / converting random data
+//! 
+//! The `RngCore` trait allows generators to implement a common interface for
+//! retrieving random data, but how should you use this? Typically users should
+//! use the `Rng` trait not `RngCore`; this provides more flexible ways to
+//! access the same data (e.g. `gen()` can output many more types than
+//! `next_u32()` and `next_u64()`; Rust's optimiser should eliminate any
+//! overhead). It also provides several useful algorithms,
+//! e.g. `gen_bool(p)` to generate events with weighted probability and
+//! `shuffle(&mut v[..])` to randomly-order a vector.
+//! 
+//! The `distributions` module provides several more ways to convert random
+//! data to useful values, e.g. time of decay is often modelled with an
+//! exponential distribution, and the log-normal distribution provides a good
+//! model of many natural phenomona.
+//! 
+//! The `seq` module has a few tools applicable to sliceable or iterable data.
+//! 
 //! # Cryptographic security
```

Comment:
Your new paragraphs are a good improvement, but I don't like it completely yet. It reads as very serious (can be a good thing), but Rand is not really a library for cryptography. I would mostly point away the responsibility to specialized crates, when one wants to start using an RNG for encrypting and stuff.

What I would like to see somewhere (just noting from my lazy chair) is that for many algorithms and situations it is good to be unpredictable. That is what I see as *the* reason for including CSPRNGs. We should make it that they are not only a good idea for some corner of problems, but are a good choice for many situations that are seemingly security-unrelated. Most things that are somehow using an RNG in a situation that can be influenced by users, do well using one. One example is the standard library's HashMap, another something like random-pivot quicksort from the [PCG blog](http://www.pcg-random.org/posts/on-trivial-predictability.html).

Reply:
Rand is not intended to be used for encryption directly (although there isn't much reason various generators couldn't be used as stream ciphers; just initialisation isn't standard), but I think Rand should be usable as a source of secrets — with caveats that (for now at least) the code hasn't had proper expert review and that currently we have limited means to mitigate side-channel attacks (which may or may not be improved upon).

Reply:
BTW while this isn't intent to be a guide to cryptography I feel there are so many myths surrounding cryptography that it is worth trying to put people straight (e.g. distrust of CSPRNGs; the bigger problem is usually *user-space* PRNGs or poorly seeded PRNGs).